src/catalog/src/stash.rs

```diff
@@ -50,148 +52,295 @@ use crate::transaction::{
+        retry_open(stash, now, bootstrap_args, None).await
+    }
+
+    #[tracing::instrument(name = "storage::open", level = "info", skip_all)]
+    async fn open(
+        &mut self,
         now: NowFn,
         bootstrap_args: &BootstrapArgs,
         deploy_generation: Option<u64>,
-    ) -> Result<Connection, (Stash, Error)> {
-        // Initialize the Stash if it hasn't been already
-        let mut conn = if !match stash.is_initialized().await {
-            Ok(is_init) => is_init,
-            Err(e) => {
-                return Err((stash, e.into()));
-            }
-        } {
-            // Get the current timestamp so we can record when we booted. We don't have to worry
-            // about `boot_ts` being less than a previously used timestamp because the stash is
-            // uninitialized and there are no previous timestamps.
-            let boot_ts = now();
-
-            // Initialize the Stash
-            let args = bootstrap_args.clone();
-            match stash
-                .with_transaction(move |mut tx| {
-                    Box::pin(async move {
-                        initialize::initialize(&mut tx, &args, boot_ts, deploy_generation).await
-                    })
-                })
-                .await
-            {
-                Ok(()) => {}
-                Err(e) => return Err((stash, e.into())),
-            }
+    ) -> Result<Connection, Error> {
+        self.open_stash().await?;
+        let stash = self.stash.take().expect("opened above");
+        retry_open(stash, now, bootstrap_args, deploy_generation).await
+    }
 
-            Connection { stash }
-        } else {
-            // Before we do anything with the Stash, we need to run any pending upgrades and
-            // initialize new collections.
-            if !stash.is_readonly() {
-                match stash.upgrade().await {
-                    Ok(()) => {}
-                    Err(e) => {
-                        return Err((stash, e.into()));
-                    }
-                }
-            }
+    async fn is_initialized(&mut self) -> Result<bool, Error> {
+        let stash = match &mut self.stash {
+            None => match self.open_stash_read_only().await {
+                Ok(stash) => stash,
+                Err(e) if e.can_recover_with_write_mode() => return Ok(false),
+                Err(e) => return Err(e.into()),
+            },
+            Some(stash) => stash,
+        };
+        stash.is_initialized().await.err_into()
+    }
 
-            let mut conn = Connection { stash };
+    async fn get_deployment_generation(&mut self) -> Result<Option<u64>, Error> {
+        let stash = match &mut self.stash {
+            None => match self.open_stash_read_only().await {
+                Ok(stash) => stash,
+                Err(e) if e.can_recover_with_write_mode() => return Ok(None),
+                Err(e) => return Err(e.into()),
+            },
```

Comment:
The code looks relatively harmless, but this appears uncovered

Reply:
I added some tests that should cover this.